block-producer/src/state_view/mod.rs

```diff
@@ -180,10 +180,16 @@ fn ensure_tx_inputs_constraints(
     match tx_inputs.account_hash {
         Some(store_account_hash) => {
             if candidate_tx.initial_account_hash() != store_account_hash {
-                return Err(VerifyTxError::IncorrectAccountInitialHash {
-                    tx_initial_account_hash: candidate_tx.initial_account_hash(),
-                    store_account_hash: Some(store_account_hash),
-                });
+                // Prevents this check from returning an Err() in case a new account
+                // that has not yet been stored in the DB submits a transaction. By default
+                // the `store_account_hash` of a new account would be `Digest::default()`
+                if store_account_hash == Digest::default() {
+                } else {
+                    return Err(VerifyTxError::IncorrectAccountInitialHash {
+                        tx_initial_account_hash: candidate_tx.initial_account_hash(),
+                        store_account_hash: Some(store_account_hash),
+                    });
+                }
             }
         },
         None => {
```

Comment:
After looking at this a bit more, a few thoughts:

- It seems like there is some inconsistency here between how store and block producer work. Specifically, if the account is not found in the store, `account_hash` should be `None` rather than `Digest::default()`. So, `Digest::default()` case should never happen and we should not handle it in a special way.
- If the account is not present in the store, this means that the transaction was executed against a new account. Ideally, we should handle this case separately. Specifically, `ProvenTransaction` should have a flag indicating whether it was executed against a new account or not, and this flag should be enforced in the transaction kernel. I'll create an issue for this in `miden-base`.

Overall, this section should look something like this:
```Rust
match tx_inputs.account_hash {
    Some(store_account_hash) => {
        if candidate_tx.initial_account_hash() != store_account_hash {
            return Err(VerifyTxError::IncorrectAccountInitialHash {
                tx_initial_account_hash: candidate_tx.initial_account_hash(),
                store_account_hash: Some(store_account_hash),
            });
        }
    },
    None => {
        // check if the transaction was executed against new account
    }
}
```

Reply:
My understanding of the situation is the following: 

1. In the miden-node we are calling the `get_tx_inputs` function on the `Store` trait that has been redefined in the Block Producer crate. 

https://github.com/0xPolygonMiden/miden-node/blob/3f657a0f45ab22b36382c16446bdd66b4d74715f/block-producer/src/store/mod.rs#L117-L192

2. This function sends a request to the API of the Miden Store. Which makes a call to the `get_tx_inputs` function of the `State` of the Miden Store.

https://github.com/0xPolygonMiden/miden-node/blob/3f657a0f45ab22b36382c16446bdd66b4d74715f/store/src/state.rs#L488-L541

3. This is the function responsible for building the response which is:

```rust
Result<(AccountState, Vec<NullifierStateForTransactionInput>), anyhow::Error>
```

4. This response is then used in the `Block Producer` to build the `TxInputs` struct. The part that we care about in the construction of this response is the `account_hash` of the `AccountState`: 

https://github.com/0xPolygonMiden/miden-node/blob/3f657a0f45ab22b36382c16446bdd66b4d74715f/store/src/state.rs#L499-L503

5. This `account_hash` is built from the `account_tree` using the `account_id`. the `open` method of the `account_tree` returns a `ValuePath` where we isolate the `value`

https://github.com/0xPolygonMiden/crypto/blob/c86bdc6d515046513a86ec331da5a0f77fd29cf2/src/merkle/path.rs#L164-L171

6. We then use into to turn this `value` which is an `RpoDigest` into a `Word`.

We understand here that the information that we receive in the Miden Node by calling `get_tx_inputs` is a `TxInputs` built through this process. 

Which also clarifies the fact that there is no possibility here for a Digest to be `None` considering that the `account_hash` that is built is an `RpoDigest` turned into a `Word`.

Reply:
I have the following questions: 

What is the behaviour that we should have in the Miden Store and in the Miden Node?

Should the `account_tree` return `None` if the account does not exist in the tree?

Reply:
> Should the `account_tree` return `None` if the account does not exist in the tree?

I don't think we need to modify the `account_tree` (I believe we use `Smt` for this).

I guess my point was more about `TxInputs` struct and how it is built from the store response. Basically, it look like this right now:
```Rust
pub struct TxInputs {
    pub account_hash: Option<Digest>,
    pub nullifiers: BTreeMap<Digest, bool>,
}
```
If we use `Digest::default()` to indicate that the account is not present in the store, then `account_hash` shouldn't be an option. But I think it would actually better if it is `Option` and we initialize it appropriately based on the store's response (when we construct `TxInputs`).

Reply:
An `Option<Digest>` is already what we have:

https://github.com/0xPolygonMiden/miden-node/blob/3f657a0f45ab22b36382c16446bdd66b4d74715f/block-producer/src/store/mod.rs#L52-L60

What I understand then is that we do not need to change the Store but we should change the way that we build our `TxInputs` in the Node:

https://github.com/0xPolygonMiden/miden-node/blob/3f657a0f45ab22b36382c16446bdd66b4d74715f/block-producer/src/store/mod.rs#L144-L164

Hence adding the fact that if `account_hash` == `Digest::default()` then we return `None`, does that seem correct to you?

Reply:
> Hence adding the fact that if `account_hash` == `Digest::default()` then we return `None`, does that seem correct to you?

Yes - i think that would be the correct way to do it.

Reply:
> After looking at this a bit more, a few thoughts:
> 
> * It seems like there is some inconsistency here between how store and block producer work. Specifically, if the account is not found in the store, `account_hash` should be `None` rather than `Digest::default()`. So, `Digest::default()` case should never happen and we should not handle it in a special way.
> * If the account is not present in the store, this means that the transaction was executed against a new account. Ideally, we should handle this case separately. Specifically, `ProvenTransaction` should have a flag indicating whether it was executed against a new account or not, and this flag should be enforced in the transaction kernel. I'll create an issue for this in `miden-base`.
> 
> Overall, this section should look something like this:
> 
> ```rust
> match tx_inputs.account_hash {
>     Some(store_account_hash) => {
>         if candidate_tx.initial_account_hash() != store_account_hash {
>             return Err(VerifyTxError::IncorrectAccountInitialHash {
>                 tx_initial_account_hash: candidate_tx.initial_account_hash(),
>                 store_account_hash: store_account_hash,
>             });
>         }
>     },
>     None => {
>         // check if the transaction was executed against new account
>     }
> }
> ```

Could you clarify here what you meant by: `// check if the transaction was executed against new account`?

Reply:
For now, this would be just a `TODO`. Once we implement the transaction kernel change I mentioned above (I still need to create an issue for this), we'll come back and handle it properly.